### PR TITLE
feat(burrow): implement core concept burrow

### PR DESCRIPTION
**Pull request by ggbot123**
## Description

添加功能：新建洞、废弃洞、查看某个洞、查看我的地洞、查看我的关注（洞）、查看我的收藏（帖子）
## How Has This Been Tested

测试：发送请求，检查返回信息是否正确

## Screenshots(optional)

1. 查看某个洞
请求：GET，burrow_id = 1，携带token，url：/burrows/<burrow_id>
返回：
图片

2. 新建洞
请求：POST，自动生成json，含title、description字段，携带token，url：/burrows/
返回：
图片

3. 废弃洞
请求1：DEL，burrow_id = 1，携带token，url：/burrows/<burrow_id>
返回：
图片

查看用户拥有的的洞，可见确实没有了burrow_id为1的洞
图片

请求2：DEL，burrow_id = 100， 携带token
返回：
图片

请求3：DEL，burrow_id = 4（不属于该用户），携带token
返回：
图片

4. 查看用户收藏的帖子
请求：GET，page = 0，携带token，url：/users/collection
返回：
图片

5. 查看用户关注的洞
请求：GET，携带token，url：/users/follow
返回：
图片

6. 查看用户拥有的洞
请求：GET，携带token，url：/users/burrow
返回：
图片

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] Code follows the code style of this project.
- [x] Changes follow the **CONTRIBUTING** guidelines.
- [x] Update necessary documentation accordingly.
- [x] Lint and tests pass locally with the changes.
- [x] Check issues and pull requests first. You don't want to duplicate effort.

### CODE REVIEW

File: backend/src/routes/burrow.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+                };
+                let burrows_id = [valid_burrows.clone(), banned_burrows.clone()].concat();
+                if burrows_id.contains(&burrow_id) {
+                    // update valid_burrow / banned_burrow in user_status table
+                    let mut ac_state: pgdb::user_status::ActiveModel = state.into();
+                    // do some type-convert things, and fill in the row according to different situations
+                    if valid_burrows.contains(&burrow_id) {
+                        valid_burrows.remove(valid_burrows.binary_search(&burrow_id).unwrap());
+                        let valid_burrows: Vec<String> =
+                            valid_burrows.iter().map(|x| x.to_string()).collect();
```
Comment (BobAnkh) [inline backend/src/routes/burrow.rs:174]: Maybe i64 can be joined directly?

**Review by BobAnkh — CHANGES_REQUESTED** [review]
Also resolve cargo clippy warnings.

I will review users interfaces later

File: backend/src/routes/burrow.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+                        let mut valid_burrows_str = valid_burrows.join(",") + ",";
+                        if valid_burrows_str == "," {
+                            valid_burrows_str = "".to_string();
+                        }
```
[the remark this answers is not in the document]
Comment (ggbot123) [reply backend/src/routes/burrow.rs:181]: Here's the reason: I kinda need "," at the end of string to call `get_valid_burrow` util, so I add an additional comma after fn `join`, while the comma won't be needed  when string is empty, so in that case I throw it off by adding this
(I know it's sort of cumbersome)

File: backend/src/utils/get_valid_burrow.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            Some(user) => {
+                let mut vec_str: Vec<&str> = user.banned_burrow.split(',').collect();
+                match vec_str.pop() {
+                    Some(_) => Ok(vec_str.iter().map(|x| x.parse::<i64>().unwrap()).collect()),
+                    None => Ok(Vec::new()),
+                }
+            }
+            None => Err("User not found.".to_string()),
+        },
+        Err(e) => Err("Database Error:".to_string() + &e.to_string()),
```
[the remark this answers is not in the document]
Comment (ggbot123) [reply backend/src/utils/get_valid_burrow.rs]: I've updated this interface and dump these sentences already

File: backend/src/routes/user.rs [hunk 4]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+                        error!(
+                            "[GET BURROW] Failed to get valid burrows: {:?}",
+                            e.to_string()
+                        );
+                        return (Status::InternalServerError, Json(Vec::new()));
+                    }
+                };
+                let burrows_id = [valid_burrows, banned_burrows].concat();
+                let mut response = Vec::new();
+                for burrow_id in burrows_id {
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:363]: Do not use same var name in for-loop, this may cause problem

File: backend/src/routes/user.rs [hunk 5]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+                        Ok(opt_burrow) => match opt_burrow {
+                            Some(burrow) => {
+                                response.push(UserGetBurrowResponse {
+                                    id: burrow.burrow_id,
+                                    title: burrow.title.clone(),
+                                    description: burrow.description.clone(),
+                                    post_num: burrow.post_num,
+                                });
+                            }
+                            None => {
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:377]: Use map.

File: backend/src/routes/user.rs [hunk 6]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+) -> (Status, Json<Vec<UserGetFavResponse>>) {
+    let pg_con = db.into_inner();
+    match pgdb::user_like::Entity::find()
+        .filter(pgdb::user_like::Column::Uid.eq(sso.id))
+        .all(&pg_con)
+        .await
+    {
+        Ok(results) => {
+            let mut posts = Vec::new();
+            for result in &results {
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:415]: use map

File: backend/src/routes/user.rs [hunk 7]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+
+#[get("/favorite")]
+pub async fn get_favorite(
+    db: Connection<PgDb>,
+    sso: sso::SsoAuth,
+) -> (Status, Json<Vec<UserGetFavResponse>>) {
+    let pg_con = db.into_inner();
+    match pgdb::user_like::Entity::find()
+        .filter(pgdb::user_like::Column::Uid.eq(sso.id))
+        .all(&pg_con)
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:410]: use paginator. frontend should give you a param of page in uri(query string). See content post for usage examples of paginator.

File: backend/src/routes/user.rs [hunk 8]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+    db: Connection<PgDb>,
+    sso: sso::SsoAuth,
+) -> (Status, Json<String>) {
+    let pg_con = db.into_inner();
+    let uid = sso.id;
+    let user_like = pgdb::user_like::ActiveModel {
+        uid: Set(uid),
+        post_id: Set(post_id),
+    };
+    match user_like.insert(&pg_con).await {
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:563]: They can be removed, for we have a interface for it.

File: backend/src/routes/user.rs [hunk 9]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+                                    title: x.title.clone(),
+                                    description: x.description.clone(),
+                                    // TODO
+                                    update: false,
+                                })
+                                .collect(),
+                        ),
+                    )
+                }
+                Err(e) => {
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:509]: Don't forget to change them.

File: backend/src/routes/user.rs [hunk 10]
```diff
@@ -286,3 +317,261 @@ pub async fn user_log_in(
...
+        },
+        Err(e) => {
+            error!("[GET BURROW] Database Error: {:?}", e.to_string());
+            return (Status::InternalServerError, Json(Vec::new()));
+        }
+    }
+}
+
+#[get("/favorite")]
+pub async fn get_favorite(
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:403]: Better change this to collection.

We will return what user collection has, but don't return what user likes.

And I don't think we need to return burrow?

**Review by BobAnkh — CHANGES_REQUESTED** [review]

File: backend/src/routes/user.rs (hunk 10 above)
Comment (ggbot123) [reply backend/src/routes/user.rs:403]: Here I returned as the doc says. Might be changed later

Comment (BobAnkh) [reply backend/src/routes/user.rs:403]: copy to have a collection interface either maybe.

File: backend/src/routes/user.rs [hunk 11]
```diff
@@ -104,29 +149,41 @@ pub async fn user_sign_up(
...
+            ..Default::default()
+        };
+        // insert rows in database
+        // <Fn, A, B> -> Result<A, B>
+        match pg_con
+            .transaction::<_, (), DbErr>(|txn| {
+                Box::pin(async move {
+                    users.insert(txn).await?;
+                    users_status.insert(txn).await?;
+                    Ok(())
```
Comment (BobAnkh) [inline backend/src/routes/user.rs:174]: Should also generate a new burrow of a new user

**Pull request merged**